// schema-util/lib/types.js
module.exports = {
  defaultType: 'string',
  supported: [
    'String', // 字符串
    'Image', // 图片
    'Number', // 数字
    'Date', // 日期
    'URL', // url
    'Boolean', // 布尔值
    'RichText', // 富文本
    'Text', // 文本
    'Color', // 颜色
    'Enum', // 下拉选择
    'Component', // 组件数据
    'Box', // 盒子，对于运营为一个可拖拽区域，对于开发者是一组css样式
    'File', // html 文件
    'Resource', // 通用自定义类型
    'Plugin', // 插件
  ],
  // 各个类型支持的 meta 字段
  meta: {
    array: ['minLength', 'maxLength', 'rule', 'sourceType', 'reuseable', 'showKey', 'if'],
    object: ['rule', 'reuseable', 'if'],
    string: [
      'minLength', 'maxLength', 'regExp', 'errorMsg',
      'rule', 'default', 'defaultValue', 'description', 'if',
    ],
    image: [
      'width', 'height', 'maxWidth', 'maxHeight', 'regExp',
      'errorMsg', 'rule', 'default', 'defaultValue',
      'setDefaultColorTo', 'description', 'if', 'maxSize'
    ],
    number: [
      'step', 'min', 'max', 'rule', 'default',
      'defaultValue', 'description', 'if',
    ],
    boolean: [
      'rule', 'default', 'defaultValue', 'description', 'if',
    ],
    url: ['regExp', 'errorMsg', 'rule', 'default', 'defaultValue', 'description', 'if'],
    text: ['regExp', 'errorMsg', 'rule', 'default', 'defaultValue', 'description', 'if'],
    richtext: ['rule', 'default', 'defaultValue', 'description', 'if'],
    color: ['rule', 'default', 'defaultValue', 'description', 'if'],
    enum: ['rule', 'default', 'defaultValue', 'description', 'source', 'if', 'multiple'],
    box: [],
    file: ['rule', 'default', 'defaultValue', 'description', 'if'],
    date: ['rule', 'default', 'defaultValue', 'description', 'if'],
    resource: ['default', 'defaultValue', 'description', 'if', 'type'],
    plugin: ['type', 'description', 'if', 'default', 'defaultValue' ],
  }
};
