// webpack-dev-middleware/dist/utils/compatibleAPI.js

function getHeaderNames(res) {
  if (typeof res.getHeaderNames !== "function") {
    return Object.keys(res._headers || {});
  }
  return res.getHeaderNames();
}

function getHeaderFromRequest(req, name) {
  if (typeof req.get === "function") {
    return req.get(name)
  }
  return req.headers[name];
}

function getHeaderFromResponse(res, name) {
  if (typeof res.get === "function") {
    return res.get(name)
  }
  return res.getHeader(name);
}

function setHeaderForResponse(res, name, value) {
  if (typeof res.set === "function") {
    res.set(name, typeof value === "number" ? String(value) : value);
    return;
  } 
  res.setHeader(name, value);
}

function setStatusCode(res, code) {
  if (typeof res.status === "function") {
    res.status(code);
    return;
  } 
  res.statusCode = code;
}

function send(req, res, bufferOtStream, byteLength) {
  if (typeof bufferOtStream.pipe === "function") {
    setHeaderForResponse(res, "Content-Length", byteLength);
    if (req.method === "HEAD") {
      res.end();
      return;
    }
    bufferOtStream.pipe(res);
    return;
  }

  if (typeof res.send === "function") {
    res.send(bufferOtStream);
    return;
  } 
  res.setHeader("Content-Length", byteLength);

  if (req.method === "HEAD") {
    res.end();
  } else {
    res.end(bufferOtStream);
  }
}

module.exports = {
  getHeaderNames,
  getHeaderFromRequest,
  getHeaderFromResponse,
  setHeaderForResponse,
  setStatusCode,
  send
};
